### PR TITLE
Factor CMake Code further

### PR DESCRIPTION
**Pull request by mgaudet**
This PR ~~(built atop #1425, so will be rebased after that gets merged. Hence WIP)~~ starts pursuing a 'data-driven' model for configuring our CMake builds. This is very similar to the approach the compiler component has been using in Make for a while. 

The idea here is to try to dramatically reduce the amount of conditional code, by instead providing services and variables that are customized to the situation at hand.

### CODE REVIEW

**Comment by charliegracie** [comment]
@rwy0717 and @dnakamura

File: cmake/platform/toolcfg/gnu.cmake [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#      The Apache License v2.0 is available at
+#      http://www.opensource.org/licenses/apache2.0.php
+#
+# Contributors:
+#    Multiple authors (IBM Corp.) - initial implementation and documentation
+###############################################################################
+
+
+
+set(OMR_WARNING_AS_ERROR_FLAG -Wall)
```
Comment (dnakamura) [inline cmake/platform/toolcfg/gnu.cmake]: what you actually want here is `-Werror`

Comment (mgaudet) [inline cmake/platform/toolcfg/gnu.cmake]: Whoops. Willfix.

Comment (dnakamura) [inline cmake/platform/toolcfg/gnu.cmake]: that being said. We probably want to add `-Wall` to all the gcc builds. Regardless of whether `OMR_WARNINGS_AS_ERRORS` is set

Comment (mgaudet) [inline cmake/platform/toolcfg/gnu.cmake]: Fixed the `-Wall` != `-Werror` issue. As for adding `-Wall` everywhere else, I'd like to do that in a subsequent PR.

**Comment by mgaudet** [comment]
In some offline discussions with @rwy0717 he took a few issues with the name Toolconfig. I'm willing to rename it, and we've sort of agreed that maybe the way to go is to call it a 'Toolchain Config' (`toolchain_config`). However, that's something I'll tackle in a subsequent PR.

**Comment by mgaudet** [comment]
Rebased this today, and assuming CI passes, is AFAIK good to go.

File: cmake/platform/arch/x86.cmake [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#    Multiple authors (IBM Corp.) - initial implementation and documentation
+###############################################################################
+
+set(OMR_ARCH_DEFINITIONS "")
+
+if(OMR_ENV_DATA64)
+   list(APPEND OMR_ARCH_DEFINITIONS J9HAMMER)
+else()
+   list(APPEND OMR_ARCH_DEFINITIONS J9X86)
+endif()
```
Comment (mgaudet) [inline cmake/platform/arch/x86.cmake]: I think the overlap between the two is sufficiently large that I'd like to keep them in one file unless we have some evidence that we should split them out. Previous experience in compiler builds indicated they were substantially the same.

File: cmake/DetectSystemInformation.cmake [hunk 3]
```diff
@@ -26,21 +26,25 @@
...
+      set(OMR_HOST_ARCH "arm")
       set(OMR_ARCH_ARM ON)
       set(OMR_ENV_DATA32 ON)
       set(OMR_TARGET_DATASIZE 32)
       set(OMR_ENV_LITTLE_ENDIAN ON)
    elseif(CMAKE_SYSTEM_NAME MATCHES "OS390")
+      set(OMR_HOST_ARCH "z")
       set(OMR_ARCH_S390 ON)
       set(OMR_ENV_DATA32 ON)
       set(OMR_TARGET_DATASIZE 32)
```
Comment (mgaudet) [inline cmake/DetectSystemInformation.cmake]: I think I want to keep this macro as a variable exporter that others can depend on.

File: cmake/platform/os/linux.cmake [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+#    Multiple authors (IBM Corp.) - initial implementation and documentation
+###############################################################################
+
+set(OMR_OS_DEFINITIONS 
+		LINUX
+		_FILE_OFFSET_BITS=64)
+
+set(OMR_OS_COMPILE_OPTIONS
+   -pthread
+   )
```
Comment (dnakamura) [inline cmake/platform/os/linux.cmake]: its weird as its a bit of both

File: cmake/platform/toolcfg/verify.cmake [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      message(FATAL_ERROR "Missing variable definition ${var} from toolconfig ${OMR_TOOLCONFIG}")
+   endif()
+endmacro(omr_assert_defined)
+
+# Warn if a variable isn't defined
+macro(omr_warn_if_not_defined var) 
+   if(NOT DEFINED ${var}) 
+      message(WARNING "Missing variable definition ${var} from toolconfig ${OMR_TOOLCONFIG}")
+   endif()
+endmacro(omr_warn_if_not_defined)
```
Comment (rwy0717) [inline cmake/platform/toolcfg/verify.cmake]: PR #1458 updated, and should suit your needs.

File: cmake/platform/os/win.cmake [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+   _WINSOCKAPI_
+   _WIN32_WINVER=${OMR_WINVER}
+   _WIN32_WINNT=${OMR_WINVER}
+   _DLL
+   _HAS_EXCEPTIONS=0
+   )
+
+if(OMR_ENV_DATA64)
+   list(APPEND OMR_OS_DEFINITIONS 
+      WIN64
```
Comment (mgaudet) [inline cmake/platform/os/win.cmake]:
```
omrmakefiles/rules.win.mk
27:  GLOBAL_CPPFLAGS+=-D_AMD64_=1 -DWIN64 -D_WIN64 -DWIN32 -D_WIN32 -DJ9HAMMER
````

Seems this is how it was happening in the Makefiles as well.

**Comment by mgaudet** [comment]
Format corrections are done, and prefixed `add_prefix` with `omr_`. 

I'd like this to get in so we can clean up more of it in subsequent PRs.

**Comment by rwy0717** [comment]
looks good to me.

File: scripts/build-on-travis.sh [hunk 7]
```diff
@@ -26,7 +26,7 @@ if test "x$BUILD_WITH_CMAKE" = "xyes"; then
   cd build
   time cmake -Wdev -GNinja -C../cmake/caches/Travis.cmake ..
   if test "x$RUN_BUILD" != "xno"; then
-    time ninja -j1  
+    time ninja -j$JOBS
```
Comment (charliegracie) [inline scripts/build-on-travis.sh]: @dnakamura didn't you have an issue with ninja files and parallel compiles in the past?

**Pull request merged**